### PR TITLE
Support custom error pages for production SSR errors in react-server

### PR DESCRIPTION
**Pull request by ismail-syed**
## Description

Related to https://github.com/Shopify/quilt/issues/854

This PR enables `@shopify/react-server` applications to provide custom error pages which we can now use on production SSR errors. It does this by: 

* Exposing a `renderError` option in `react-server`'s `createServer`. `renderError` is similar to the standard `render` option we pass, except it's for rendering a consumer provided `Error` component. If `renderError` is not set, then react-server defaults to returning a static HTML error page as a sane default/fallback. 
* `react-server-webpack-plugin` has also been updated to utilizes the new `renderError` API in `react-server.` The plugin checks for the existence of an `error` component. If an `error` component exists, it gets imported in the server and passed to `react-server`. This also creates a virtual client entry point for the `error` component for client-side hydration. 

## 🎩 

1. checkout this branch and run `yarn run build`
2. `cd packages/react-server && yarn link`
3. `cd packages/react-server-webpack-plugin && yarn link` 
4. checkout the `quilt_rails-prod-ssr-error` branch in `shrink-ray` 
5. `yarn link @shopify/react-server && yarn link @shopify/react-server-webpack-plugin`
6.  Run the following to run the rails server in production mode 

```
$ NODE_ENV=production SK_SIMULATE_PRODUCTION=1 bundle exec rake assets:precompile
$ NODE_ENV=production SK_SIMULATE_PRODUCTION=1 dev run
```
7. in a second window run `node build/server/main.js`
8. Open the app and verify that you see the following error page. This is an error page setup in the `shrink-ray` branch. 

Screen Shot 2020-06-04 at 3 15 05 AM

10. Delete the error component file: `app/ui/error.tsx`
11. Re-due steps 6, 7, 8 
12. Verify that you see the following screenshot. This is the default fallback error page provided by `react-server`. 

Screen Shot 2020-06-04 at 1 28 25 PM

## Type of change

- [ ]  Patch: Bug (non-breaking change which fixes an issue)
- [x]  Minor: New feature (non-breaking change which adds functionality)
- [ ]  Major: Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have added a changelog entry, prefixed by the type of change noted above (Documentation fix and Test update does not need a changelog as we do not publish new version)

### CODE REVIEW

File: packages/react-server/src/render/error/fallback-error-markup.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                          We received an error from the server. Please check
+                          again later.
+                        </p>
+                      </div>
+                    </div>
+                  </div>
+                </div>
+                <div class="Polaris-EmptyState__ImageContainer">
+                  <img
+                    src="https://cdn.shopify.com/s/files/1/0757/9955/files/empty-state.svg"
```
Comment (ismail-syed) [inline packages/react-server/src/render/error/fallback-error-markup.ts:54]: Note, this is the not right empty state image I plan to use for the fallback error page. 

**wrong one:**
image
 

**right one:**
image


### Why aren't you using the right one?

I put the expected svg in a [google cloud storage](https://console.cloud.google.com/storage/browser/cdn.shopifycloud.com/web-foundation/images). Getting this image publicly accessible is WIP.

File: packages/react-server/src/render/error/fallback-error-markup.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
+export const fallbackErrorMarkup = `
```
Comment (TheMallen) [inline packages/react-server/src/render/error/fallback-error-markup.ts:1]: Might be nicer to have this as an actual template, but I know the pathing would be a bit of a pain.

**Comment by michenly** [comment]
I try 🎩  and it running into a strange error where my browser would download the document instead of opening it in the browser. I tried both Chrome and Firefox and its the same behaviour. Also checked that this does not occur with shrink-ray master.

Screen Shot 2020-06-04 at 11 41 23 AM

File: packages/react-server/src/render/error/fallback-error-markup.ts (hunk 1 above)
Comment (ismail-syed) [reply packages/react-server/src/render/error/fallback-error-markup.ts]: PR has been updated

File: packages/react-server-webpack-plugin/src/react-server-webpack-plugin.ts [hunk 3]
```diff
@@ -129,7 +153,7 @@ function clientSource() {
   `;
 }
 
-function noSourceExists(
+export function noSourceExists(
```
[the remark this answers is not in the document]
Comment (ismail-syed) [reply packages/react-server-webpack-plugin/src/react-server-webpack-plugin.ts:156]: Good catch!

File: packages/react-server/src/render/error/fallback-error-markup.ts (hunk 2 above)
Comment (ismail-syed) [reply packages/react-server/src/render/error/fallback-error-markup.ts:1]: yah, i felt that pain and went with this approach instead

**Comment by ismail-syed** [comment]
Tested and verified in shrink-ray. Merging and cutting a release.

**Pull request merged**